### PR TITLE
Gravity projection imu dev

### PR DESCRIPTION
**Pull request by aikolina**
I modified `inverse-kinematics` in order to be able to project a 3D point along a direction specified by IMU's measurements.

In particular, I added the following methods:
* `setProjectionDirection` defines the direction along which we want to project a 3D point (see `InverseKinematics.h/cpp)`,
* `setPimu` defines the projection matrix according to the IMU's measurements (see `ConvexHullHelpers.h/cpp`),
* `projectAlongGravity` projects a 3D point according to the projection matrix defined by `setPimu`  (see `ConvexHullHelpers.h/cpp`).

The method `projectAlongGravity` substitutes the method `project` in:
* `getCenterOfMassProjectionMargin` in `InverseKinematics.cpp`,
* `updateState` in `InverseKinematicsNLP.cpp`.

### CODE REVIEW

File: src/inverse-kinematics/src/ConvexHullHelpers.cpp [hunk 1]
```diff
@@ -342,5 +342,20 @@ namespace iDynTree
...
     }
 
+    Vector2 ConvexHullProjectionConstraint::projectAlongGravity(iDynTree::Position& posIn3dInAbsoluteFrame)
+    {
+        iDynTree::Vector2 projected;
+        toEigen(projected) = toEigen(Pimu)*toEigen(posIn3dInAbsoluteFrame-o);
+        return projected;
+    }
+
+    void ConvexHullProjectionConstraint::setPimu(const iDynTree::Direction xProjection, const iDynTree::Direction yProjection, const iDynTree::Position originOfPlaneInWorld)
```
Comment (traversaro) [inline src/inverse-kinematics/src/ConvexHullHelpers.cpp]: This method is meant to set the rows of the projection matrix, that are not constained to be orthonormal. However, the `iDynTree::Direction` class is meant to represent direction in the 3D space, and so its content are forced to be normalized (see https://github.com/robotology/idyntree/blob/2470a71cee49ee3a611195e36b8dc447eca9f154/src/core/src/Direction.cpp#L23). 
This will create deformation in the projection logic. 
To avoid this problem, I suggest to modify the API to remove the `Pimu` attribute and the `setPimu` method and just have a single method `ConvexHullProjectionConstraint::projectAlongDirection` that takes the direction along which to project the com everytime, and will compute the Pimu on the fly.

**Comment by traversaro** [comment]
See https://github.com/robotology/idyntree/pull/353/commits/45c0989c127da0b5389731cd571370ef45a90e19  for the skeleton of a unit test (check over method that are not existing are still commented) using the API I was thinking of.

File: src/inverse-kinematics/src/ConvexHullHelpers.cpp [hunk 2]
```diff
@@ -342,20 +342,31 @@ namespace iDynTree
         return margin;
     }
 
-    Vector2 ConvexHullProjectionConstraint::projectAlongGravity(iDynTree::Position& posIn3dInAbsoluteFrame)
+    void ConvexHullProjectionConstraint::setProjectionAlongDirection(Vector3 direction, const iDynTree::Position originOfPlaneInWorld)
```
Comment (traversaro) [inline src/inverse-kinematics/src/ConvexHullHelpers.cpp]: @aikolina  The `originOfPlaneInWorld` argument seems to be unused.

File: src/inverse-kinematics/tests/ConvexHullHelpersUnitTest.cpp [hunk 3]
```diff
@@ -97,19 +97,37 @@ void testConvexHullProjectionWithGravity()
...
+    // If the direction along witch we are projecting is the normal of the plane, the projection
     // should match the one done without direction
-    iDynTree::Direction planeNormal(0.0, 0.0, -1.0);
+
+    //iDynTree::Direction planeNormal(0.0, 0.0, -1.0);
     //ASSERT_EQUAL_VECTOR(projectionConstraint.project(testPoint), projectionConstraint.projectAlongDirection(testPoint, planeNormal));
+    iDynTree::Vector3 planeNormalVector;
+    planeNormalVector.setVal(0, 0.0);
+    planeNormalVector.setVal(1, 0.0);
+    planeNormalVector.setVal(2, -1.0);
```
Comment (traversaro) [inline src/inverse-kinematics/tests/ConvexHullHelpersUnitTest.cpp]: This is indeed a direction, so it is ok to use `iDynTree::Direction` for it.

File: src/inverse-kinematics/tests/ConvexHullHelpersUnitTest.cpp [hunk 4]
```diff
@@ -97,19 +97,37 @@ void testConvexHullProjectionWithGravity()
...
+
+    projectionConstraint.setProjectionAlongDirection(planeNormalVector, iDynTree::Position::Zero());
+
+    ASSERT_EQUAL_VECTOR(projectionConstraint.project(testPoint), projectionConstraint.projectAlongDirection(testPoint));
+
+    //ASSERT_IS_TRUE(false);
+
+    // If the direction of the projection is "skewed" to the left, the point should instead be inside the convex hull
+    //iDynTree::Direction skewedDirection(-1.0, 0.0, -1.0);
+    //ASSERT_IS_TRUE(projectionConstraint.computeMargin(projectionConstraint.projectAlongDirection(testPoint, skewedDirection)) < 0);
```
Comment (traversaro) [inline src/inverse-kinematics/tests/ConvexHullHelpersUnitTest.cpp]: If we went for the other API, we can remove all this comments.

File: src/inverse-kinematics/tests/ConvexHullHelpersUnitTest.cpp [hunk 5]
```diff
@@ -97,19 +97,37 @@ void testConvexHullProjectionWithGravity()
...
+    iDynTree::Vector3 skewedDirection;
+    skewedDirection.setVal(0, -1.0);
+    skewedDirection.setVal(1, 0.0);
+    skewedDirection.setVal(2, 1.0);
```
Comment (traversaro) [inline src/inverse-kinematics/tests/ConvexHullHelpersUnitTest.cpp]: I think a `projectionConstraint.setProjectionAlongDirection(skewedDirection, iDynTree::Position::Zero())` is missing here (and this is one of the reason for which a single call `projectAlongDirection(point,direction)` could be more handy.

Comment (aikolina) [reply src/inverse-kinematics/tests/ConvexHullHelpersUnitTest.cpp]: Yes, I forgot it. Thanks!

File: src/inverse-kinematics/tests/ConvexHullHelpersUnitTest.cpp (hunk 3 above)
Comment (aikolina) [reply src/inverse-kinematics/tests/ConvexHullHelpersUnitTest.cpp]: I didn't notice that the class `Direction` derives from `Vector3`. Fixed!

File: src/inverse-kinematics/src/ConvexHullHelpers.cpp (hunk 2 above)
Comment (aikolina) [reply src/inverse-kinematics/src/ConvexHullHelpers.cpp]: I forgot to add a line:
```C++
o = originOfPlaneInWorld;
```

File: src/inverse-kinematics/tests/ConvexHullHelpersUnitTest.cpp (hunk 5 above)
Comment (aikolina) [reply src/inverse-kinematics/tests/ConvexHullHelpersUnitTest.cpp]: So would it be better to store the `direction` vector inside `InverseKinematics`?

But considering the possible uses of the method `projectAlongDirection`, I think it's more logic to keep it associated with the class `ConvexHullProjectionConstraint`. What do you think? @traversaro

Comment (traversaro) [reply src/inverse-kinematics/tests/ConvexHullHelpersUnitTest.cpp]: I think the API would be easier to use by `projectAlongDirection(point,direction)`, and for any use in `InverseKinematics` it can always be called by passing the projection direction saved in `InverseKinematics` . In this case we add an attribute in `InverseKinematics`, but eventually we will need to refactor that logic in its own class, so I think it is reasonable to do so.

File: src/inverse-kinematics/src/ConvexHullHelpers.cpp (hunk 2 above)
Comment (traversaro) [reply src/inverse-kinematics/src/ConvexHullHelpers.cpp]: I am a bit confused why you should change the origin of the plane on which you are projecting the point in this call. If this is the case, why don't we set again also `xAxisOfPlaneInWorld` and `yAxisOfPlaneInWorld` ?

File: src/inverse-kinematics/src/ConvexHullHelpers.cpp [hunk 6]
```diff
@@ -342,5 +342,33 @@ namespace iDynTree
...
     }
 
+    void ConvexHullProjectionConstraint::setProjectionAlongDirection(Vector3 direction, const iDynTree::Position originOfPlaneInWorld)
+    {
+        o = originOfPlaneInWorld;
+
+        Vector3 xProjection, yProjection;
+
+        // define the projection for the x-component
+        xProjection.setVal(0, 1.0);
```
Comment (traversaro) [inline src/inverse-kinematics/src/ConvexHullHelpers.cpp]: I just realized that in this expression we are assuming that `xAxisPlaneInWorld` is equal to `(1, 0, 0)` and `yAxisPlaneInWorld` is `(0, 1, 0)`. For the time being it is ok to keep like this until we fully debug all the problems, but we should at least document it in the method documentation.

File: src/tools/idyntree-sole-gui/plugin/ObserverThread.cpp [hunk 7]
```diff
@@ -323,12 +387,23 @@ void ObserverThread::threadRelease()
...
+    {
+        m_imuPort.interrupt();
+        yarp::os::Network::connect(m_remoteImuPortName.c_str(),
+                                   m_localImuPortName.c_str());
+        m_imuPort.close();
+    }
+
     skin_port.close();
+
+    delete filtIMUGravity;
```
Comment (traversaro) [inline src/tools/idyntree-sole-gui/plugin/ObserverThread.cpp]: filtIMUGravity = 0;

File: src/inverse-kinematics/src/InverseKinematicsNLP.cpp [hunk 8]
```diff
@@ -238,8 +238,16 @@ namespace kinematics {
...
+//                 iDynTree::toEigen(comInfo.projectedCom) =
+//                 iDynTree::toEigen(m_data.m_comHullConstraint.P)*iDynTree::toEigen(comInfo.com - m_data.m_comHullConstraint.o);
+
+                // NOTE: The previous lines could be substituted by:
+                // comInfo.projectedCom = m_data.m_comHullConstraint.project(comInfo.com);
+
+                // Project the COM along the gravity vector
+                comInfo.projectedCom = m_data.m_comHullConstraint.projectAlongDirection(comInfo.com);
+//                std::cerr << "COM: " << comInfo.com.getVal(0) << ", " << comInfo.com.getVal(1) << ", " << comInfo.com.getVal(2) << std::endl;
+//                std::cerr << "Projected COM: " << comInfo.projectedCom.getVal(0) << ", " << comInfo.projectedCom.getVal(1) << std::endl << std::endl;
```
Comment (traversaro) [inline src/inverse-kinematics/src/InverseKinematicsNLP.cpp]: Cleanup the comments.

File: src/inverse-kinematics/src/InverseKinematics.cpp [hunk 9]
```diff
@@ -624,6 +625,15 @@ namespace iDynTree {
...
     }
 
+    void InverseKinematics::setProjectionDirection(iDynTree::Vector3 direction,
+                                                   const iDynTree::Position originOfPlaneInWorld)
+    {
+        // define the projection matrix 'Pdirection' in the class 'ConvexHullProjectionConstraint'
+        IK_PIMPL(m_pimpl)->m_comHullConstraint.setProjectionAlongDirection(direction, originOfPlaneInWorld);
+
+//        // TODO: set the projection direction from a given one
+//        projectionDirection = direction;
```
Comment (traversaro) [inline src/inverse-kinematics/src/InverseKinematics.cpp]: Remove?

File: src/inverse-kinematics/include/iDynTree/InverseKinematics.h [hunk 10]
```diff
@@ -705,6 +705,16 @@ class iDynTree::InverseKinematics
...
 
+    /*!
+     * Set the directions along which a point will be projected.
+     *
+     * \author Aiko Dinale (29/08/2017)
+     *
+     * @param direction     vector along which we want to project a point
+     */
+    void setProjectionDirection(iDynTree::Vector3 direction,
+                                const iDynTree::Position originOfPlaneInWorld = iDynTree::Position::Zero());
```
Comment (traversaro) [inline src/inverse-kinematics/include/iDynTree/InverseKinematics.h]: We are modifying the origin set in the ConvexHull here. We should keep the origin that is set while the convex hull is built.

File: src/inverse-kinematics/src/ConvexHullHelpers.cpp [hunk 11]
```diff
@@ -327,8 +327,18 @@ namespace iDynTree
             {
                 distanceWithoutSign = candidateDistance;
             }
+//             std::cerr << "Vertex " << i << "/" << projectedConvexHull.getNrOfVertices() << " = " << projectedConvexHull(i).getVal(0)
+                                                                                        << ", " << projectedConvexHull(i).getVal(1) << std::endl;
```
Comment (traversaro) [inline src/inverse-kinematics/src/ConvexHullHelpers.cpp]: Cleanup.

File: src/inverse-kinematics/include/iDynTree/InverseKinematics.h (hunk 10 above)
Comment (traversaro) [reply src/inverse-kinematics/include/iDynTree/InverseKinematics.h]: The name of the method is a bit confusing. I would suggest `setCOMConstraintProjectionDirection` to clarify the projection that we are referencing.

File: src/tests/CMakeLists.txt [hunk 12]
```diff
@@ -20,10 +20,10 @@ endif()
...
 if(IDYNTREE_USES_ICUB_MAIN)
     # Add benchmark test against yarp/icub functions
-    add_subdirectory(yarp_benchmark)
+    #add_subdirectory(yarp_benchmark)
 
     # Add a subdirectory of all the consistency tests between the old
     # iCub-dependent external force/internal torque estimation
     #  implementations and the new native iDynTree implementations
-    add_subdirectory(icub_consistency)
+    #add_subdirectory(icub_consistency)
```
Comment (traversaro) [inline src/tests/CMakeLists.txt]: We should revert this changes are rebase this branch on the top of the `devel` branch.

File: src/inverse-kinematics/include/iDynTree/InverseKinematics.h [hunk 13]
```diff
@@ -640,7 +640,7 @@ class iDynTree::InverseKinematics
 
     /*! @name Solution-related methods
       */
-    ///@{
```
Comment (traversaro) [inline src/inverse-kinematics/include/iDynTree/InverseKinematics.h]: This removal is breaking the documentation group at https://github.com/robotology/idyntree/pull/353/files#diff-c85a932fafd5e5f39278aee4923d544eL669 .

File: src/inverse-kinematics/src/InverseKinematics.cpp (hunk 9 above)
Comment (aikolina) [reply src/inverse-kinematics/src/InverseKinematics.cpp]: Removed.

File: src/inverse-kinematics/src/InverseKinematicsNLP.cpp (hunk 8 above)
Comment (aikolina) [reply src/inverse-kinematics/src/InverseKinematicsNLP.cpp]: Done.

File: src/inverse-kinematics/src/ConvexHullHelpers.cpp (hunk 11 above)
Comment (aikolina) [reply src/inverse-kinematics/src/ConvexHullHelpers.cpp]: Done.

File: src/inverse-kinematics/include/iDynTree/InverseKinematics.h (hunk 13 above)
Comment (aikolina) [reply src/inverse-kinematics/include/iDynTree/InverseKinematics.h]: Fixed.

File: src/inverse-kinematics/include/iDynTree/InverseKinematics.h (hunk 10 above)
Comment (aikolina) [reply src/inverse-kinematics/include/iDynTree/InverseKinematics.h]: You're definitely right, no idea of why I did so. Now the method is simply:
```C++
void setCOMConstraintProjectionDirection(iDynTree::Vector3 direction);
```

File: src/tools/idyntree-sole-gui/plugin/ObserverThread.cpp (hunk 7 above)
Comment (aikolina) [reply src/tools/idyntree-sole-gui/plugin/ObserverThread.cpp]: Done.

File: src/tools/idyntree-sole-gui/plugin/ObserverThread.cpp [hunk 14]
```diff
@@ -393,16 +468,58 @@ void ObserverThread::draw(QPainter* qpainter, int widthInPixels, int heightInPix
     pen.setColor(QColor("green"));
     qpainter->setPen(pen);
 
-    double safetyMarginX=0.01;
-    double safetyMarginY=0.005;
-    iDynTree::Polygon primarySolePolygonInPrimarySole = iDynTree::Polygon::XYRectangleFromOffsets(0.025, 0.015, 0.01, 0.01);
-    iDynTree::Polygon secondarySolePolygonInSecondarySole = iDynTree::Polygon::XYRectangleFromOffsets(0.025, 0.015, 0.01, 0.01);
-    iDynTree::Polygon primarySolePolygonWithSafetyInPrimarySole  = iDynTree::Polygon::XYRectangleFromOffsets(0.025-safetyMarginX, 0.015-safetyMarginX, 0.01-safetyMarginY, 0.01-safetyMarginY);
-    iDynTree::Polygon secondarySolePolygonWithSafetyInSecondarySole = iDynTree::Polygon::XYRectangleFromOffsets(0.025-safetyMarginX, 0.015-safetyMarginX, 0.01-safetyMarginY, 0.01-safetyMarginY);
+    double safetyMarginX = 0.012;
```
[the remark this answers is not in the document]
Comment (aikolina) [reply src/tools/idyntree-sole-gui/plugin/ObserverThread.cpp]: Done.

File: src/tests/CMakeLists.txt (hunk 12 above)
Comment (aikolina) [reply src/tests/CMakeLists.txt]: Uncommented.

File: src/tools/idyntree-sole-gui/plugin/ObserverThread.cpp [hunk 15]
```diff
@@ -311,6 +338,43 @@ void ObserverThread::run()
...
+            yarp::sig::Vector gravityDirectionInSensorFrameFilteredV;
+            gravityDirectionInSensorFrameFilteredV = filtIMUGravity->filt(gravityDirectionInSensorFrameV);
+
+            iDynTree::Direction gravityDirectionInSensorFrameFiltered;
+            gravityDirectionInSensorFrameFiltered(0) = gravityDirectionInSensorFrameFilteredV[0];
+            gravityDirectionInSensorFrameFiltered(1) = gravityDirectionInSensorFrameFilteredV[1];
+            gravityDirectionInSensorFrameFiltered(2) = gravityDirectionInSensorFrameFilteredV[2];
+
+//             gravityDirectionInSensorFrame.Normalize();
+//
```
[the remark this answers is not in the document]
Comment (aikolina) [reply src/tools/idyntree-sole-gui/plugin/ObserverThread.cpp]: Removed.

File: src/inverse-kinematics/src/InverseKinematics.cpp [hunk 16]
```diff
@@ -289,6 +291,17 @@ namespace iDynTree {
...
+	// Initialize the COM's projection direction in such a way that is along the lien perpendicular to the xy-axes of the World
+	iDynTree::Direction zAxis;
+	toEigen(zAxis) = toEigen(xAxisOfPlaneInWorld).cross(toEigen(yAxisOfPlaneInWorld));
+	
+	Vector3 zAxisOfPlaneInWorld;
+	zAxisOfPlaneInWorld(0) = zAxis.getVal(0);
+	zAxisOfPlaneInWorld(1) = zAxis.getVal(1);
+	zAxisOfPlaneInWorld(2) = zAxis.getVal(2);
+		
+	IK_PIMPL(m_pimpl)->m_comHullConstraint.setProjectionAlongDirection(zAxisOfPlaneInWorld);
```
Comment (traversaro) [inline src/inverse-kinematics/src/InverseKinematics.cpp]: I think we can directly call `IK_PIMPL(m_pimpl)->m_comHullConstraint.setProjectionAlongDirection(zAxis)` here.

File: src/inverse-kinematics/src/ConvexHullHelpers.cpp [hunk 17]
```diff
@@ -339,8 +341,35 @@ namespace iDynTree
...
     }
 
+    void ConvexHullProjectionConstraint::setProjectionAlongDirection(Vector3 direction)
+    {
+        Vector3 xProjection, yProjection;
+
+        // define the projection for the x-component
+        xProjection.setVal(0, 1.0);
+        xProjection.setVal(1, 0.0);
+        xProjection.setVal(2, -direction.getVal(0)/direction.getVal(2));
```
Comment (francesco-romano) [inline src/inverse-kinematics/src/ConvexHullHelpers.cpp]: binary operators should have space on the left and right of the operator

File: src/inverse-kinematics/src/ConvexHullHelpers.cpp [hunk 18]
```diff
@@ -339,8 +341,35 @@ namespace iDynTree
...
+
+        // define the projection for the y-component
+        yProjection.setVal(0, 0.0);
+        yProjection.setVal(1, 1.0);
+        yProjection.setVal(2, -direction.getVal(1)/direction.getVal(2));
```
Comment (francesco-romano) [inline src/inverse-kinematics/src/ConvexHullHelpers.cpp]: see previous comment

File: src/inverse-kinematics/src/ConvexHullHelpers.cpp [hunk 19]
```diff
@@ -339,8 +341,35 @@ namespace iDynTree
...
+
+        // fill the projection matrix
+        toEigen(Pdirection).block<1,3>(0,0) = toEigen(xProjection);
```
Comment (francesco-romano) [inline src/inverse-kinematics/src/ConvexHullHelpers.cpp]: space after comma, i.e. `block<1, 3>(0, 0)`

Comment (francesco-romano) [reply src/inverse-kinematics/src/ConvexHullHelpers.cpp]: by the way, eigen should have a `row` method to access a specific row

File: src/inverse-kinematics/src/ConvexHullHelpers.cpp [hunk 20]
```diff
@@ -339,8 +341,35 @@ namespace iDynTree
...
+        toEigen(Pdirection).block<1,3>(1,0) = toEigen(yProjection);
+
+    }
+
+    Vector2 ConvexHullProjectionConstraint::projectAlongDirection(iDynTree::Position& posIn3dInAbsoluteFrame)
+    {
+        iDynTree::Vector2 projected;
+        toEigen(projected) = toEigen(Pdirection)*toEigen(posIn3dInAbsoluteFrame-o);
```
Comment (francesco-romano) [inline src/inverse-kinematics/src/ConvexHullHelpers.cpp]: space missing (two times)

File: src/inverse-kinematics/src/InverseKinematics.cpp [hunk 21]
```diff
@@ -319,7 +332,8 @@ namespace iDynTree {
         iDynTree::Position comInAbsoluteConstraintFrame =
             IK_PIMPL(m_pimpl)->m_comHullConstraint.absoluteFrame_X_supportFrame[0]*(kinDyn.getWorldTransform(IK_PIMPL(m_pimpl)->m_comHullConstraint.supportFrameIndices[0]).inverse()*kinDyn.getCenterOfMassPosition());
 
-        iDynTree::Vector2 comProjection = IK_PIMPL(m_pimpl)->m_comHullConstraint.project(comInAbsoluteConstraintFrame);
+        //iDynTree::Vector2 comProjection = IK_PIMPL(m_pimpl)->m_comHullConstraint.project(comInAbsoluteConstraintFrame);
```
Comment (francesco-romano) [inline src/inverse-kinematics/src/InverseKinematics.cpp]: if a line is not needed, you can remove it instead of commenting it

File: src/inverse-kinematics/src/InverseKinematicsNLP.cpp [hunk 22]
```diff
@@ -238,8 +238,10 @@ namespace kinematics {
 
             if (m_data.m_comHullConstraint.isActive()) {
                 // Project the jacobian and the com
-                iDynTree::toEigen(comInfo.projectedCom) =
-                    iDynTree::toEigen(m_data.m_comHullConstraint.P) * iDynTree::toEigen(comInfo.com - m_data.m_comHullConstraint.o);
+                // comInfo.projectedCom = m_data.m_comHullConstraint.project(comInfo.com);
```
Comment (francesco-romano) [inline src/inverse-kinematics/src/InverseKinematicsNLP.cpp]: same as previous comment

File: src/inverse-kinematics/src/InverseKinematicsNLP.cpp [hunk 23]
```diff
@@ -933,7 +935,7 @@ namespace kinematics {
                 if (m_data.m_comHullConstraint.isActive()) {
                     // Project the jacobian
                     iDynTree::toEigen(comInfo.projectedComJacobian) =
-                            iDynTree::toEigen(m_data.m_comHullConstraint.AtimesP) * iDynTree::toEigen(comInfo.comJacobianAnalytical);
+                            iDynTree::toEigen(m_data.m_comHullConstraint.A)* iDynTree::toEigen(m_data.m_comHullConstraint.Pdirection) * iDynTree::toEigen(comInfo.comJacobianAnalytical);
```
Comment (francesco-romano) [inline src/inverse-kinematics/src/InverseKinematicsNLP.cpp]: space

**Comment by traversaro** [comment]
Tests are working fine ( https://travis-ci.org/robotology/idyntree/jobs/272953724 ) except for an issue that has been solved in `devel` by this commit: https://github.com/robotology/idyntree/commit/852d42e8c025f3a79ac758502b64017b4820f396 . Merging!
